### ISSUE TITLE
Linux /proc/self/exe is not reliable for getting snapshot inside self-contained executable

### ISSUE THREAD

**Issue by ntkme** [dart-lang/sdk#50926]
When executing a self-contained executable (dartaotruntime + aot-snapshot), the way we get snapshot is first `File::ReadLink("/proc/self/exe")`, and then attempt to read snapshot from the file path as following: 

https://github.com/dart-lang/sdk/blob/8e7b1119b3461de82b27fc92fd6821b6b84c9ac8/runtime/bin/main.cc#L1190-L1191

https://github.com/dart-lang/sdk/blob/24683da915867609b9b52bd2124e6fb603b9f160/runtime/bin/platform_linux.cc#L157-L163

However, `/proc/self/exe` is not a reliable way to get the path to current executable. It can be broken in a few cases, but in particular, if the executable is explicitly launched via `ld-linux`. Here is a demonstration:

```
$ /usr/bin/readlink -f /proc/self/exe
/usr/bin/readlink
$ /lib64/ld-linux-x86-64.so.2 /usr/bin/readlink -f /proc/self/exe
/usr/lib/x86_64-linux-gnu/ld-2.31.so
```

The behavior of dartaotruntime is affected by this, that when launching via `ld-linux`, the call to `Snapshot::TryReadAppendedAppSnapshotElf(executable_path)` will effectively read `ld-linux` and not be able to correct locate the embedded aot-snapshot inside the elf binary, for which it will simply behave as running a normal dart vm instead of running the aot snapshot.

This is the root cause of: https://github.com/sass/dart-sass-embedded/issues/78. Interestingly, it seems on certain linux distribution or with certain config, this issue is present even if the user is directly launching the executable without explicitly invoking `ld-linux` from command line. In such case, it is a bigger issue as all self-contained executables are broken on such platform.

**Comment by mraleph** [comment]
@ntkme Do you actually have a reproduction for the issue outside of loading through runtime linker directly? Because if that's the only case when it breaks - I would be tempted to mark this low priority. The real fix here is to start doing real linking and stop doing concatenation of files, which we have been using as a stop gap solution.

**Comment by ntkme** [comment]
Personally I don’t have a way to reproduce it other than invoking through the linker, but it did happen for one user when invoking directly: https://github.com/sass/dart-sass-embedded/issues/78#issuecomment-1065433430

Unfortunately, that user stopped replying the thread so that we don’t know much about the environment and configuration where it happened.

**Comment by ntkme** [comment]
Another way to break this is use `patchelf` to modify the interpreter, which is common for NixOS users. Technically it is a different issue but closely related. As patchelf don’t understand the magic offset, once the binary is patch the offset would be wrong: https://github.com/ntkme/sass-embedded-host-ruby/pull/91#issuecomment-1373000531

**Comment by mraleph** [comment]
Yeah, the issue with `patchelf` has the same underlying reason - doing real linking (or at the very least real ELF sections editing) is the answer.

**Comment by NANASHI0X74** [comment]
I tried to package a dart executable for nixos and apparently I've run into this exact issue.

**Comment by NANASHI0X74** [comment]
@mraleph you said that the concatenating of aot snapshots was a stopgap solution, but are there any plans or ongoing efforts to do the real linking as you suggested? 😄

**Comment by mraleph** [comment]
No ongoing effort at the moment, sorry. If you want to distribute AOT compiled Dart code to the OS where our current approach breaks then your next best option currently is to distribute AOT snapshot and AOT runtime as separate binaries and have a shell script which wraps these two things together.
